File: 2022/lib/day02.dart
part of 'aoc.dart';

class Day02 extends AOC {
  @override
  int part1() {
    int score = 0;
    for (var element in data) {
      final data = element.split('');
      String a = data.first;
      String b = data.last;
      if (b == 'X') {
        score += 1;
      } else if (b == 'Y') {
        score += 2;
      } else if (b == 'Z') {
        score += 3;
      }
      if (a == 'A' && b == 'Y') {
        score += 6;
      } else if (a == 'B' && b == 'Z') {
        score += 6;
      } else if (a == 'C' && b == 'X') {
        score += 6;
      } else if (a == 'A' && b == 'X') {
        score += 3;
      } else if (a == 'B' && b == 'Y') {
        score += 3;
      } else if (a == 'C' && b == 'Z') {
        score += 3;
      }
    }
    return score;
  }

  @override
  int part2() {
    int score = 0;
    for (var element in data) {
      final data = element.split('');
      String a = data.first;
      String b = data.last;
      if (b == 'X') {
        score += 0;
      } else if (b == 'Y') {
        score += 3;
      } else if (b == 'Z') {
        score += 6;
      }
      if (a == 'A') {
        switch (b) {
          case 'Y':
            score += 1;
            break;
          case 'Z':
            score += 2;
            break;
          case 'X':
            score += 3;
            break;
        }
      } else if (a == 'B') {
        switch (b) {
          case 'X':
            score += 1;
            break;
          case 'Y':
            score += 2;
            break;
          case 'Z':
            score += 3;
            break;
        }
      } else if (a == 'C') {
        switch (b) {
          case 'Z':
            score += 1;
            break;
          case 'X':
            score += 2;
            break;
          case 'Y':
            score += 3;
            break;
        }
      }
    }
    return score;
  }
}

File: 2024/Sources/Day/Day02.swift
//
//  Day02.swift
//  AOC
//
//  Created by Jeevan Chandra Joshi on 02/12/24.
//

import Foundation

struct Day02: Day {
    let data: String

    var lines: [[Int]] {
        data.split(separator: "\n").map {
            $0.split(separator: " ").compactMap { Int($0) }
        }
    }

    func isValid(_ line: [Int]) -> Bool {
        let isIncreasing = line.first! < line.last!

        for i in 0 ..< line.count - 1 {
            let diff = isIncreasing ? (line[i + 1] - line[i]) : (line[i] - line[i + 1])
            if diff < 1 || diff > 3 {
                return false
            }
        }

        return true
    }

    func part1() -> Int {
        lines.reduce(0) { sum, line in
            if isValid(line) {
                return sum + 1
            }
            return sum + 0
        }
    }

    func part2() -> Int {
        lines.reduce(0) { sum, line in
            if isValid(line) {
                return sum + 1
            }
            for i in 0 ..< line.count {
                var line = line
                line.remove(at: i)
                if isValid(line) {
                    return sum + 1
                }
            }
            return sum + 0
        }
    }
}
